src/fs/bundle.rs

```diff
@@ -0,0 +1,235 @@
+use crate::fs::{FsNode, GuestPath};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::fmt::Debug;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use zip::result::ZipError;
+use zip::ZipArchive;
+
+/// A helper struct to build an FsNode with files and directories coming in arbitrary order.
+/// This is required, because ZIP files are allowed to store entries in arbitrary order.
+struct FsNodeBuilder {
+    root: FsNode,
+}
+
+impl FsNodeBuilder {
+    fn new() -> Self {
+        Self {
+            root: FsNode::dir(),
+        }
+    }
+
+    fn find_or_make_directory(&mut self, path: &GuestPath) -> &mut FsNode {
+        let mut current = &mut self.root;
+        for part in path.as_str().split('/') {
+            if part.is_empty() {
+                continue;
+            }
+            let FsNode::Directory { children, .. } = current else {
+                panic!("expected directory, got {:?}", current);
+            };
+
+            let next = children
+                .entry(part.to_string())
+                .or_insert_with(|| FsNode::dir());
+            current = next;
+        }
+        current
+    }
+
+    fn add_file(&mut self, path: &GuestPath, node: FsNode) {
+        // I __hope__ no adversory will try to create files with .. in their names...
+        let (parent_name, file_name) = path.parent_and_file_name().unwrap();
+        let dir = self.find_or_make_directory(parent_name);
+        let FsNode::Directory { children, .. } = dir else {
+            panic!("expected directory, got {:?}", dir);
+        };
+
+        children.insert(file_name.to_string(), node);
+    }
+
+    fn add_directory(&mut self, path: &GuestPath) {
+        self.find_or_make_directory(path);
+    }
+
+    fn build(self) -> FsNode {
+        self.root
+    }
+}
+
+/// Represents an open app bundle, either a directory or a zip file.
+pub enum BundleData {
+    HostDirectory(PathBuf),
+    Zip {
+        zip: ZipArchive<std::fs::File>,
+        /// Path to the app bundle inside the zip file.
+        /// It should be "Payload/<app name>.app" (no trailing slash!).
+        bundle_path: String,
+    },
+}
+
+impl BundleData {
+    // TODO actually, the error type is not at all a ZipError.. I would like to use a custom error type here (anyhow, snafu, thiserror, ...), but there doesnt seem to be one used in the project?
+    fn find_bundle_path_in_archive(
+        zip: &mut ZipArchive<std::fs::File>,
+    ) -> Result<String, ZipError> {
+        static BUNDLE_NAME_REGEX: Lazy<Regex> =
+            Lazy::new(|| Regex::new(r"Payload/([^/]+)\.app/").unwrap());
```

Comment:
It's not a big deal, but I think we could match this pattern without needing a regex library.

Reply:
I don't want to reinvent regex =(

Reply:
That's… fair. I just think Rust usually offers enough nice ways to do things (e.g. the `str` methods and `Iter`) that regex is a bit overkill, and I'm a bit wary of it being deceptively simple sometimes (e.g. unclear greedy vs non-greedy matching behaviour).

I think something like this would work?

```rust
if let Some(name) = path
    .strip_prefix("Payload/")
    .and_then(|path| path.split_once('/'))
    .and_then(|(name, _)| name.strip_suffix(".app")) {
```